client/virtual_channel_settlement.go

```diff
@@ -0,0 +1,188 @@
+	}
+
+	virtualBalsRemapped := virtual.translateBalances(virtualAlloc.IndexMap)
+
+	// We assume that the asset types of parent channel and virtual channel are the same.
+	state.Allocation.Balances = state.Allocation.Balances.Add(virtualBalsRemapped)
+
+	if err := state.Allocation.RemoveSubAlloc(virtualAlloc); err != nil {
+		c.Log().WithError(err).Panicf("removing sub-allocation with id %x", virtualAlloc.ID)
+	}
+
+	err = c.updateGeneric(ctx, state, func(mcu *msgChannelUpdate) wire.Msg {
+		return &virtualChannelSettlementProposal{
+			msgChannelUpdate:     *mcu,
+			ChannelParams:        *virtual.Params(),
+			FinalState:           *virtual.state(),
+			FinalStateSignatures: virtual.machine.CurrentTX().Sigs,
+		}
+	})
+
+	return errors.WithMessage(err, "update parent channel")
+}
+
+func (c *Client) handleVirtualChannelSettlementProposal(
+	parent *Channel,
+	prop *virtualChannelSettlementProposal,
+	responder *UpdateResponder,
+) {
+	err := c.validateVirtualChannelSettlementProposal(parent, prop)
+	if err != nil {
+		c.rejectProposal(responder, err.Error())
+	}
+
+	ctx, cancel := context.WithTimeout(c.Ctx(), virtualChannelFundingTimeout)
+	defer cancel()
+
+	err = c.settlementWatcher.Await(ctx, prop)
+	if err != nil {
+		c.rejectProposal(responder, err.Error())
+	}
+
+	c.acceptProposal(responder)
+}
+
+func (c *Client) validateVirtualChannelSettlementProposal(
+	parent *Channel,
+	prop *virtualChannelSettlementProposal,
+) error {
+	// Validate parameters.
+	if prop.ChannelParams.ID() != prop.FinalState.ID {
+		return errors.New("invalid parameters")
+	}
+
+	// Validate signatures.
+	for i, sig := range prop.FinalStateSignatures {
+		ok, err := channel.Verify(
+			prop.ChannelParams.Parts[i],
+			&prop.ChannelParams,
+			&prop.FinalState,
+			sig,
+		)
+		if err != nil {
+			return err
+		} else if !ok {
+			return errors.New("invalid signature")
+		}
+	}
+
+	// Validate allocation.
+
+	// Assert equal assets.
+	if err := channel.AssetsAssertEqual(parent.state().Assets, prop.FinalState.Assets); err != nil {
+		return errors.WithMessage(err, "assets do not match")
+	}
+
+	// Assert contained before and matching funds
+	subAlloc, containedBefore := parent.state().SubAlloc(prop.ChannelParams.ID())
+	if !containedBefore || !subAlloc.BalancesEqual(prop.FinalState.Sum()) {
+		return errors.New("virtual channel not allocated")
+	}
+
+	// Assert not contained after
+	_, containedAfter := prop.State.SubAlloc(prop.ChannelParams.ID())
+	if containedAfter {
+		return errors.New("virtual channel must not be de-allocated after update")
+	}
+
+	// Assert correct balances
+	virtual := transformBalances(prop.FinalState.Balances, parent.state().NumParts(), subAlloc.IndexMap)
+	correctBalances := parent.state().Balances.Add(virtual).Equal(prop.State.Balances)
+	if !correctBalances {
+		return errors.New("invalid balances")
+	}
+
+	return nil
+}
+
+func (c *Client) matchSettlementProposal(a, b interface{}) (ok bool) {
+	// Cast.
```

Comment:
Everywhere where we `return false` (or just after line 155), it might be worth logging with `c.Log().Warnf`, because it either indicates cheating peers or some bug, right? Right now, we just `return false` and the problems or `err` errors go unnoticed.

Reply:
I think this also applies to the other matcher function

Reply:
You could also do something like
```go
var err error
defer func() {
  if ok == false {
    c.Log().Warnf("Error in matching settlement proposal: %+v", err)
  }
}()
```
and then set the `err` variable to a descriptive error if we don't get one already from a function call.